{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "81fe3e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "from io import StringIO\n",
    "from time import sleep\n",
    "import re\n",
    "import time\n",
    "\n",
    "from pathlib import Path\n",
    "from typing import Iterable, Dict, Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "139438d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDBID</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Family</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>RES</th>\n",
       "      <th>Text</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5bz3</td>\n",
       "      <td>Na(+)/H(+) antiporter</td>\n",
       "      <td>Sodium-proton antiporter 1</td>\n",
       "      <td>Q5SIA2</td>\n",
       "      <td>2.30</td>\n",
       "      <td>CRYSTAL STRUCTURE OF SODIUM PROTON ANTIPORTER ...</td>\n",
       "      <td>Thermus thermophilus (strain HB8 / ATCC 27634 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4qtn</td>\n",
       "      <td>Nicotinamide riboside transporter PnuC</td>\n",
       "      <td>Nicotinamide mononucleotide transporter</td>\n",
       "      <td>D2ZZC1</td>\n",
       "      <td>2.80</td>\n",
       "      <td>Crystal structure of the Vitamin B3 transporte...</td>\n",
       "      <td>Neisseria mucosa ATCC 25996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5ezm</td>\n",
       "      <td>4-amino-4-deoxy-L-arabinose transferase or rel...</td>\n",
       "      <td>Aminoarabinose transferase family</td>\n",
       "      <td>Q1LDT6</td>\n",
       "      <td>2.70</td>\n",
       "      <td>Crystal Structure of ArnT from Cupriavidus met...</td>\n",
       "      <td>Cupriavidus metallidurans (strain ATCC 43123 /...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4od4</td>\n",
       "      <td>4-hydroxybenzoate octaprenyltransferase</td>\n",
       "      <td>UbiA prenyltransferases</td>\n",
       "      <td>Q9YBM8</td>\n",
       "      <td>3.30</td>\n",
       "      <td>Apo structure of a UbiA homolog from Aeropyrum...</td>\n",
       "      <td>Aeropyrum pernix (strain ATCC 700893 / DSM 118...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4p19</td>\n",
       "      <td>425aa long hypothetical proton glutamate sympo...</td>\n",
       "      <td>Proton glutamate symport protein</td>\n",
       "      <td>O59010</td>\n",
       "      <td>3.25</td>\n",
       "      <td>Closed, apo inward-facing state of the glutama...</td>\n",
       "      <td>Pyrococcus horikoshii (strain ATCC 700860 / DS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>3vou</td>\n",
       "      <td>Voltage-gated sodium channel</td>\n",
       "      <td>Voltage-gated ion channels</td>\n",
       "      <td>A3SUL8</td>\n",
       "      <td>3.20</td>\n",
       "      <td>The crystal structure of NaK-NavSulP chimera c...</td>\n",
       "      <td>Sulfitobacter sp. NAS-14.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>4g7y</td>\n",
       "      <td>Voltage-sensor containing phosphatase</td>\n",
       "      <td>Voltage-sensor domain</td>\n",
       "      <td>Q4W8A1</td>\n",
       "      <td>2.80</td>\n",
       "      <td>Crystal structure of voltage sensing domain of...</td>\n",
       "      <td>Ciona intestinalis (Transparent sea squirt) (A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>5c78</td>\n",
       "      <td>WlaB protein</td>\n",
       "      <td>ABC transporter B family (ABCB)</td>\n",
       "      <td>O86150</td>\n",
       "      <td>2.90</td>\n",
       "      <td>ATP-driven lipid-linked oligosaccharide flippa...</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>3ddl</td>\n",
       "      <td>Xanthorhodopsin</td>\n",
       "      <td>Ion-translocating microbial rhodopsin</td>\n",
       "      <td>Q2S2F8</td>\n",
       "      <td>1.90</td>\n",
       "      <td>Crystallographic Structure of Xanthorhodopsin,...</td>\n",
       "      <td>Salinibacter ruber</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>4umv</td>\n",
       "      <td>Zinc-transporting ATPase</td>\n",
       "      <td>P-ATPase</td>\n",
       "      <td>Q3YW59</td>\n",
       "      <td>3.20</td>\n",
       "      <td>CRYSTAL STRUCTURE OF A ZINC-TRANSPORTING PIB-T...</td>\n",
       "      <td>Shigella sonnei (strain Ss046)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    PDBID                                       Protein Name  \\\n",
       "0    5bz3                              Na(+)/H(+) antiporter   \n",
       "1    4qtn             Nicotinamide riboside transporter PnuC   \n",
       "2    5ezm  4-amino-4-deoxy-L-arabinose transferase or rel...   \n",
       "3    4od4            4-hydroxybenzoate octaprenyltransferase   \n",
       "4    4p19  425aa long hypothetical proton glutamate sympo...   \n",
       "..    ...                                                ...   \n",
       "425  3vou                      Voltage-gated sodium channel    \n",
       "426  4g7y              Voltage-sensor containing phosphatase   \n",
       "427  5c78                                       WlaB protein   \n",
       "428  3ddl                                   Xanthorhodopsin    \n",
       "429  4umv                           Zinc-transporting ATPase   \n",
       "\n",
       "                                      Family UniprotID   RES  \\\n",
       "0                Sodium-proton antiporter 1     Q5SIA2  2.30   \n",
       "1    Nicotinamide mononucleotide transporter    D2ZZC1  2.80   \n",
       "2          Aminoarabinose transferase family    Q1LDT6  2.70   \n",
       "3                    UbiA prenyltransferases    Q9YBM8  3.30   \n",
       "4           Proton glutamate symport protein    O59010  3.25   \n",
       "..                                       ...       ...   ...   \n",
       "425               Voltage-gated ion channels    A3SUL8  3.20   \n",
       "426                    Voltage-sensor domain    Q4W8A1  2.80   \n",
       "427          ABC transporter B family (ABCB)    O86150  2.90   \n",
       "428    Ion-translocating microbial rhodopsin   Q2S2F8   1.90   \n",
       "429                                 P-ATPase    Q3YW59  3.20   \n",
       "\n",
       "                                                  Text  \\\n",
       "0    CRYSTAL STRUCTURE OF SODIUM PROTON ANTIPORTER ...   \n",
       "1    Crystal structure of the Vitamin B3 transporte...   \n",
       "2    Crystal Structure of ArnT from Cupriavidus met...   \n",
       "3    Apo structure of a UbiA homolog from Aeropyrum...   \n",
       "4    Closed, apo inward-facing state of the glutama...   \n",
       "..                                                 ...   \n",
       "425  The crystal structure of NaK-NavSulP chimera c...   \n",
       "426  Crystal structure of voltage sensing domain of...   \n",
       "427  ATP-driven lipid-linked oligosaccharide flippa...   \n",
       "428  Crystallographic Structure of Xanthorhodopsin,...   \n",
       "429  CRYSTAL STRUCTURE OF A ZINC-TRANSPORTING PIB-T...   \n",
       "\n",
       "                                              Organism  \n",
       "0    Thermus thermophilus (strain HB8 / ATCC 27634 ...  \n",
       "1                          Neisseria mucosa ATCC 25996  \n",
       "2    Cupriavidus metallidurans (strain ATCC 43123 /...  \n",
       "3    Aeropyrum pernix (strain ATCC 700893 / DSM 118...  \n",
       "4    Pyrococcus horikoshii (strain ATCC 700860 / DS...  \n",
       "..                                                 ...  \n",
       "425                         Sulfitobacter sp. NAS-14.1  \n",
       "426  Ciona intestinalis (Transparent sea squirt) (A...  \n",
       "427                               Campylobacter jejuni  \n",
       "428                                Salinibacter ruber   \n",
       "429                     Shigella sonnei (strain Ss046)  \n",
       "\n",
       "[430 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates_path = \"alpha_helix_candidates.txt\" # populated from HTML of website\n",
    "table_schema = [\"PDBID\", \"Protein Name\", \"Family\", \"UniprotID\", \"RES\", \"Text\", \"Organism\"]\n",
    "\n",
    "series_l = []\n",
    "with open(candidates_path, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if line.startswith(\"<tr><td>\") and line.endswith(\"</td></tr>\"):\n",
    "            line = line[8:-10]\n",
    "        else:\n",
    "            continue # poorly formatted line\n",
    "        l = line.split(\"</td><td>\")\n",
    "        d = {table_schema[i] : l[i] for i in range(len(l))}\n",
    "        series_l.append(d)\n",
    "\n",
    "df_candidates = pd.DataFrame(series_l)\n",
    "df_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66cda508",
   "metadata": {},
   "outputs": [],
   "source": [
    "UNI_BASE = \"https://rest.uniprot.org/uniprotkb/stream\"\n",
    "ACCESSION_RE = re.compile(r\"^[A-NR-Z][0-9][A-Z0-9]{3}[0-9](?:-\\d+)?$\")  # supports isoforms (P12345-2)\n",
    "\n",
    "def _normalize_uniprot_ids(series) -> list:\n",
    "    \"\"\"Split dirty cells, strip junk, validate accessions, dedupe while preserving order.\"\"\"\n",
    "    seen, clean = set(), []\n",
    "    for raw in series.fillna(\"\").astype(str):\n",
    "        for tok in re.split(r\"[^\\w\\-]+\", raw):  # split on non-word/non-dash\n",
    "            tok = tok.strip().upper().replace(\"_\", \"\")\n",
    "            if tok and ACCESSION_RE.match(tok) and tok not in seen:\n",
    "                seen.add(tok)\n",
    "                clean.append(tok)\n",
    "    return clean\n",
    "\n",
    "def _chunk(lst, n):\n",
    "    for i in range(0, len(lst), n):\n",
    "        yield lst[i:i+n]\n",
    "\n",
    "def _query_chunk(chunk, timeout=20):\n",
    "    q = \"accession:(\" + \" OR \".join(chunk) + \")\"\n",
    "    params = {\"format\": \"tsv\", \"fields\": \"accession,length\", \"query\": q}\n",
    "    r = requests.get(UNI_BASE, params=params, timeout=timeout)\n",
    "    if r.status_code == 429:\n",
    "        time.sleep(1.0)\n",
    "        r = requests.get(UNI_BASE, params=params, timeout=timeout)\n",
    "    r.raise_for_status()\n",
    "    return pd.read_csv(StringIO(r.text), sep=\"\\t\")\n",
    "\n",
    "def fetch_uniprot_lengths(uniprot_ids, chunk_size=100, pause=0.2, timeout=20):\n",
    "    \"\"\"\n",
    "    Batch-fetch sequence lengths for a list/Series of UniProt accessions.\n",
    "    Returns DataFrame with columns ['Entry','Length'].\n",
    "    Cleans/validates IDs and auto-recovers from 400 errors by shrinking chunks.\n",
    "    \"\"\"\n",
    "    ids = _normalize_uniprot_ids(pd.Series(uniprot_ids))\n",
    "    if not ids:\n",
    "        return pd.DataFrame(columns=[\"Entry\", \"Length\"])\n",
    "\n",
    "    out = []\n",
    "    try:\n",
    "        for ch in _chunk(ids, chunk_size):\n",
    "            df = _query_chunk(ch, timeout=timeout)\n",
    "            if not df.empty:\n",
    "                out.append(df[[\"Entry\", \"Length\"]])\n",
    "            if pause:\n",
    "                time.sleep(pause)\n",
    "    except requests.HTTPError as e:\n",
    "        if e.response is not None and e.response.status_code == 400:\n",
    "            # retry with smaller chunks; final fallback per-ID\n",
    "            for ch in _chunk(ids, 20):\n",
    "                try:\n",
    "                    df = _query_chunk(ch, timeout=timeout)\n",
    "                    if not df.empty:\n",
    "                        out.append(df[[\"Entry\", \"Length\"]])\n",
    "                    if pause:\n",
    "                        time.sleep(pause)\n",
    "                except requests.HTTPError as e2:\n",
    "                    if e2.response is not None and e2.response.status_code == 400:\n",
    "                        for single in ch:\n",
    "                            try:\n",
    "                                df1 = _query_chunk([single], timeout=timeout)\n",
    "                                if not df1.empty:\n",
    "                                    out.append(df1[[\"Entry\", \"Length\"]])\n",
    "                            except requests.HTTPError:\n",
    "                                pass  # skip truly bad IDs\n",
    "                    else:\n",
    "                        raise\n",
    "        else:\n",
    "            raise\n",
    "\n",
    "    return pd.concat(out, ignore_index=True) if out else pd.DataFrame(columns=[\"Entry\", \"Length\"])\n",
    "\n",
    "def add_seq_length_from_uniprot(df, overwrite_res=False, new_col_name=None):\n",
    "    \"\"\"\n",
    "    Merge UniProt sequence length into your DataFrame (by 'UniprotID').\n",
    "    - overwrite_res=True: fill/overwrite df['RES'] from UniProt length.\n",
    "    - else: create a new column ('SeqLen' or custom).\n",
    "    \"\"\"\n",
    "    lens = fetch_uniprot_lengths(df[\"UniprotID\"])\n",
    "    merged = df.merge(lens, how=\"left\", left_on=\"UniprotID\", right_on=\"Entry\").drop(columns=[\"Entry\"])\n",
    "    if overwrite_res:\n",
    "        merged[\"RES\"] = pd.to_numeric(merged.get(\"RES\"), errors=\"coerce\").fillna(merged[\"Length\"])\n",
    "        return merged.drop(columns=[\"Length\"])\n",
    "    else:\n",
    "        col = new_col_name or \"SeqLen\"\n",
    "        merged[col] = merged[\"Length\"]\n",
    "        return merged.drop(columns=[\"Length\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "829759b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDBID</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Family</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>RES</th>\n",
       "      <th>Text</th>\n",
       "      <th>Organism</th>\n",
       "      <th>SeqLen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4wav</td>\n",
       "      <td>Bacteriorhodopsin-I</td>\n",
       "      <td>Microbial and algal rhodopsins</td>\n",
       "      <td>G0LFX8</td>\n",
       "      <td>2.80</td>\n",
       "      <td>Crystal Structure of Haloquadratum walsbyi bac...</td>\n",
       "      <td>Haloquadratum walsbyi (strain DSM 16854 / JCM ...</td>\n",
       "      <td>254.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4fbz</td>\n",
       "      <td>Deltarhodopsin</td>\n",
       "      <td>Microbial and algal rhodopsins</td>\n",
       "      <td>I4DST7</td>\n",
       "      <td>2.70</td>\n",
       "      <td>Crystal structure of deltarhodopsin from Halot...</td>\n",
       "      <td>Haloterrigena thermotolerans</td>\n",
       "      <td>241.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4bem</td>\n",
       "      <td>F-type ATP synthase subunit E</td>\n",
       "      <td>V-type and F-type ATPases</td>\n",
       "      <td>H6LFT2</td>\n",
       "      <td>2.10</td>\n",
       "      <td>Crystal structure of the F-type ATP synthase c...</td>\n",
       "      <td>Acetobacterium woodii (strain ATCC 29683 / DSM...</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4tsy</td>\n",
       "      <td>Fragaceatoxin C</td>\n",
       "      <td>Anemone pore-forming cytolysin</td>\n",
       "      <td>B9W5G6</td>\n",
       "      <td>3.14</td>\n",
       "      <td>Crystal structure of FraC with lipids</td>\n",
       "      <td>Actinia fragacea (Strawberry anemone)</td>\n",
       "      <td>179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5cbg</td>\n",
       "      <td>Ion transport 2 domain protein</td>\n",
       "      <td>Two-pore K+ channel</td>\n",
       "      <td>D5UM26</td>\n",
       "      <td>3.14</td>\n",
       "      <td>Calcium activated non-selective cation channel</td>\n",
       "      <td>Tsukamurella paurometabola (strain ATCC 8368 /...</td>\n",
       "      <td>123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2oar</td>\n",
       "      <td>Large-conductance mechanosensitive channel</td>\n",
       "      <td>Large conductance mechanosensitive ion channel...</td>\n",
       "      <td>A5U127</td>\n",
       "      <td>3.50</td>\n",
       "      <td>Mechanosensitive Channel of Large Conductance ...</td>\n",
       "      <td>Mycobacterium tuberculosis.</td>\n",
       "      <td>151.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4wab</td>\n",
       "      <td>Leukotriene C4 synthase</td>\n",
       "      <td>MAPEG domain</td>\n",
       "      <td>B5MCC3</td>\n",
       "      <td>2.70</td>\n",
       "      <td>Crystal structure of mPGES1 solved by native-S...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4rng</td>\n",
       "      <td>MtN3/saliva family</td>\n",
       "      <td>SemiSWEET bacterial sugar transporters</td>\n",
       "      <td>B5YGD6</td>\n",
       "      <td>2.40</td>\n",
       "      <td>Crystal structure of a bacterial homologue of ...</td>\n",
       "      <td>Thermodesulfovibrio yellowstonii (strain ATCC ...</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5a43</td>\n",
       "      <td>Putative fluoride ion transporter CrcB</td>\n",
       "      <td>Fluoride exporter</td>\n",
       "      <td>B7LI20</td>\n",
       "      <td>2.58</td>\n",
       "      <td>Crystal structure of a dual topology fluoride ...</td>\n",
       "      <td>Escherichia coli O45:K1 (strain S88 / ExPEC)</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4hyj</td>\n",
       "      <td>Rhodopsin</td>\n",
       "      <td>Microbial and algal rhodopsins</td>\n",
       "      <td>B1YFV8</td>\n",
       "      <td>2.30</td>\n",
       "      <td>Crystal structure of Exiguobacterium sibiricum...</td>\n",
       "      <td>Exiguobacterium sibiricum (strain DSM 17290 / ...</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5ax0</td>\n",
       "      <td>Rhodopsin I</td>\n",
       "      <td>Microbial and algal rhodopsins</td>\n",
       "      <td>G3CEP6</td>\n",
       "      <td>1.52</td>\n",
       "      <td>Crystal Structure of the Cell-Free Synthesized...</td>\n",
       "      <td>Acetabularia acetabulum (Mermaid's wine glass)...</td>\n",
       "      <td>246.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3am6</td>\n",
       "      <td>Rhodopsin II</td>\n",
       "      <td>Microbial and algal rhodopsins</td>\n",
       "      <td>G1K3Q0</td>\n",
       "      <td>3.20</td>\n",
       "      <td>Crystal structure of the proton pumping rhodop...</td>\n",
       "      <td>Acetabularia acetabulum (Mermaid's wine glass)...</td>\n",
       "      <td>229.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4pop</td>\n",
       "      <td>Thiamine transporter ThiT</td>\n",
       "      <td>S component of vitamin uptake transporter</td>\n",
       "      <td>A2RI47</td>\n",
       "      <td>2.20</td>\n",
       "      <td>ThiT with LMG139 bound</td>\n",
       "      <td>Lactococcus lactis subsp. cremoris (strain MG1...</td>\n",
       "      <td>182.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4qnc</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SemiSWEET bacterial sugar transporters</td>\n",
       "      <td>B0SR19</td>\n",
       "      <td>2.39</td>\n",
       "      <td>Crystal structure of a SemiSWEET in an occlude...</td>\n",
       "      <td>Leptospira biflexa serovar Patoc (strain Patoc...</td>\n",
       "      <td>85.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4qnd</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SemiSWEET bacterial sugar transporters</td>\n",
       "      <td>F9RBV9</td>\n",
       "      <td>1.70</td>\n",
       "      <td>Crystal structure of a SemiSWEET</td>\n",
       "      <td>Vibrio sp. N418</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4xu4</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>Insig-related proteins</td>\n",
       "      <td>A1T557</td>\n",
       "      <td>1.90</td>\n",
       "      <td>Crystal structure of a mycobacterial Insig hom...</td>\n",
       "      <td>Mycobacterium vanbaalenii (strain DSM 7251 / P...</td>\n",
       "      <td>204.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PDBID                                Protein Name  \\\n",
       "0   4wav                         Bacteriorhodopsin-I   \n",
       "1   4fbz                              Deltarhodopsin   \n",
       "2   4bem               F-type ATP synthase subunit E   \n",
       "3   4tsy                             Fragaceatoxin C   \n",
       "4   5cbg              Ion transport 2 domain protein   \n",
       "5   2oar  Large-conductance mechanosensitive channel   \n",
       "6   4wab                     Leukotriene C4 synthase   \n",
       "7   4rng                          MtN3/saliva family   \n",
       "8   5a43      Putative fluoride ion transporter CrcB   \n",
       "9   4hyj                                   Rhodopsin   \n",
       "10  5ax0                                 Rhodopsin I   \n",
       "11  3am6                               Rhodopsin II    \n",
       "12  4pop                   Thiamine transporter ThiT   \n",
       "13  4qnc                     Uncharacterized protein   \n",
       "14  4qnd                     Uncharacterized protein   \n",
       "15  4xu4                     Uncharacterized protein   \n",
       "\n",
       "                                               Family UniprotID   RES  \\\n",
       "0                     Microbial and algal rhodopsins     G0LFX8  2.80   \n",
       "1                      Microbial and algal rhodopsins    I4DST7  2.70   \n",
       "2                           V-type and F-type ATPases    H6LFT2  2.10   \n",
       "3                     Anemone pore-forming cytolysin     B9W5G6  3.14   \n",
       "4                                 Two-pore K+ channel    D5UM26  3.14   \n",
       "5   Large conductance mechanosensitive ion channel...    A5U127  3.50   \n",
       "6                                        MAPEG domain    B5MCC3  2.70   \n",
       "7              SemiSWEET bacterial sugar transporters    B5YGD6  2.40   \n",
       "8                                   Fluoride exporter    B7LI20  2.58   \n",
       "9                     Microbial and algal rhodopsins     B1YFV8  2.30   \n",
       "10                     Microbial and algal rhodopsins    G3CEP6  1.52   \n",
       "11                     Microbial and algal rhodopsins    G1K3Q0  3.20   \n",
       "12          S component of vitamin uptake transporter    A2RI47  2.20   \n",
       "13             SemiSWEET bacterial sugar transporters    B0SR19  2.39   \n",
       "14             SemiSWEET bacterial sugar transporters    F9RBV9  1.70   \n",
       "15                             Insig-related proteins    A1T557  1.90   \n",
       "\n",
       "                                                 Text  \\\n",
       "0   Crystal Structure of Haloquadratum walsbyi bac...   \n",
       "1   Crystal structure of deltarhodopsin from Halot...   \n",
       "2   Crystal structure of the F-type ATP synthase c...   \n",
       "3               Crystal structure of FraC with lipids   \n",
       "4      Calcium activated non-selective cation channel   \n",
       "5   Mechanosensitive Channel of Large Conductance ...   \n",
       "6   Crystal structure of mPGES1 solved by native-S...   \n",
       "7   Crystal structure of a bacterial homologue of ...   \n",
       "8   Crystal structure of a dual topology fluoride ...   \n",
       "9   Crystal structure of Exiguobacterium sibiricum...   \n",
       "10  Crystal Structure of the Cell-Free Synthesized...   \n",
       "11  Crystal structure of the proton pumping rhodop...   \n",
       "12                             ThiT with LMG139 bound   \n",
       "13  Crystal structure of a SemiSWEET in an occlude...   \n",
       "14                   Crystal structure of a SemiSWEET   \n",
       "15  Crystal structure of a mycobacterial Insig hom...   \n",
       "\n",
       "                                             Organism  SeqLen  \n",
       "0   Haloquadratum walsbyi (strain DSM 16854 / JCM ...   254.0  \n",
       "1                        Haloterrigena thermotolerans   241.0  \n",
       "2   Acetobacterium woodii (strain ATCC 29683 / DSM...    82.0  \n",
       "3               Actinia fragacea (Strawberry anemone)   179.0  \n",
       "4   Tsukamurella paurometabola (strain ATCC 8368 /...   123.0  \n",
       "5                         Mycobacterium tuberculosis.   151.0  \n",
       "6                                Homo sapiens (Human)    93.0  \n",
       "7   Thermodesulfovibrio yellowstonii (strain ATCC ...    88.0  \n",
       "8        Escherichia coli O45:K1 (strain S88 / ExPEC)   126.0  \n",
       "9   Exiguobacterium sibiricum (strain DSM 17290 / ...   252.0  \n",
       "10  Acetabularia acetabulum (Mermaid's wine glass)...   246.0  \n",
       "11  Acetabularia acetabulum (Mermaid's wine glass)...   229.0  \n",
       "12  Lactococcus lactis subsp. cremoris (strain MG1...   182.0  \n",
       "13  Leptospira biflexa serovar Patoc (strain Patoc...    85.0  \n",
       "14                                    Vibrio sp. N418    97.0  \n",
       "15  Mycobacterium vanbaalenii (strain DSM 7251 / P...   204.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fill RES in-place from UniProt:\n",
    "df_candidates = add_seq_length_from_uniprot(df_candidates, overwrite_res=False, new_col_name=\"SeqLen\").dropna(subset=[\"SeqLen\"])\n",
    "len_limit = 256\n",
    "df_candidates = df_candidates[df_candidates[\"SeqLen\"] < len_limit].reset_index(drop=True)\n",
    "df_candidates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a523601b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_file(url, local_filename):\n",
    "    \"\"\"\n",
    "    Downloads a file from a given URL and saves it locally.\n",
    "\n",
    "    Args:\n",
    "        url (str): The URL of the file to download.\n",
    "        local_filename (str): The name to save the downloaded file as.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        response = requests.get(url, stream=True)\n",
    "        response.raise_for_status()  # Raise an exception for bad status codes (4xx or 5xx)\n",
    "\n",
    "        with open(local_filename, 'wb') as f:\n",
    "            for chunk in response.iter_content(chunk_size=8192):\n",
    "                f.write(chunk)\n",
    "        print(f\"File '{local_filename}' downloaded successfully from '{url}'\")\n",
    "\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(f\"Error downloading file: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ac1b2187",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0     4wav\n",
      "1     4fbz\n",
      "2     4bem\n",
      "3     4tsy\n",
      "4     5cbg\n",
      "5     2oar\n",
      "6     4wab\n",
      "7     4rng\n",
      "8     5a43\n",
      "9     4hyj\n",
      "10    5ax0\n",
      "11    3am6\n",
      "12    4pop\n",
      "13    4qnc\n",
      "14    4qnd\n",
      "15    4xu4\n",
      "Name: PDBID, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# now retrieve pdbs\n",
    "good_pdbids = df_candidates[\"PDBID\"]\n",
    "print(good_pdbids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b7479037",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File 'alpha_helix_pdbs/4wav.pdb' downloaded successfully from 'https://files.rcsb.org/download/4wav.pdb'\n",
      "File 'alpha_helix_pdbs/4fbz.pdb' downloaded successfully from 'https://files.rcsb.org/download/4fbz.pdb'\n",
      "File 'alpha_helix_pdbs/4bem.pdb' downloaded successfully from 'https://files.rcsb.org/download/4bem.pdb'\n",
      "File 'alpha_helix_pdbs/4tsy.pdb' downloaded successfully from 'https://files.rcsb.org/download/4tsy.pdb'\n",
      "File 'alpha_helix_pdbs/5cbg.pdb' downloaded successfully from 'https://files.rcsb.org/download/5cbg.pdb'\n",
      "File 'alpha_helix_pdbs/2oar.pdb' downloaded successfully from 'https://files.rcsb.org/download/2oar.pdb'\n",
      "File 'alpha_helix_pdbs/4wab.pdb' downloaded successfully from 'https://files.rcsb.org/download/4wab.pdb'\n",
      "File 'alpha_helix_pdbs/4rng.pdb' downloaded successfully from 'https://files.rcsb.org/download/4rng.pdb'\n",
      "File 'alpha_helix_pdbs/5a43.pdb' downloaded successfully from 'https://files.rcsb.org/download/5a43.pdb'\n",
      "File 'alpha_helix_pdbs/4hyj.pdb' downloaded successfully from 'https://files.rcsb.org/download/4hyj.pdb'\n",
      "File 'alpha_helix_pdbs/5ax0.pdb' downloaded successfully from 'https://files.rcsb.org/download/5ax0.pdb'\n",
      "File 'alpha_helix_pdbs/3am6.pdb' downloaded successfully from 'https://files.rcsb.org/download/3am6.pdb'\n",
      "File 'alpha_helix_pdbs/4pop.pdb' downloaded successfully from 'https://files.rcsb.org/download/4pop.pdb'\n",
      "File 'alpha_helix_pdbs/4qnc.pdb' downloaded successfully from 'https://files.rcsb.org/download/4qnc.pdb'\n",
      "File 'alpha_helix_pdbs/4qnd.pdb' downloaded successfully from 'https://files.rcsb.org/download/4qnd.pdb'\n",
      "File 'alpha_helix_pdbs/4xu4.pdb' downloaded successfully from 'https://files.rcsb.org/download/4xu4.pdb'\n"
     ]
    }
   ],
   "source": [
    "folder = \"alpha_helix_pdbs\"\n",
    "for id in good_pdbids:\n",
    "    download_file(f\"https://files.rcsb.org/download/{id}.pdb\", f\"{folder}/{id}.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "070ab6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_pdb(pdb_path):\n",
    "    \"\"\"\n",
    "    Parse a PDB file line by line.\n",
    "    Returns:\n",
    "      atoms:  list of dicts (serial, name, resname, chain, resseq, x,y,z)\n",
    "      helices: list of dicts (helix_id, start_res, start_chain, end_res, end_chain, length)\n",
    "    \"\"\"\n",
    "    atoms, helices = [], []\n",
    "    atom_lines = []\n",
    "    helix_lines = []\n",
    "    with open(pdb_path) as f:\n",
    "        for line in f:\n",
    "            record = line[0:6].strip()\n",
    "\n",
    "            # ---- ATOM records ----\n",
    "            if record == \"ATOM\":\n",
    "                atoms.append({\n",
    "                    \"serial\":   int(line[6:11]),\n",
    "                    \"name\":     line[12:16].strip(),\n",
    "                    \"resname\":  line[17:20].strip(),\n",
    "                    \"chain\":    line[21].strip(),\n",
    "                    \"resseq\":   int(line[22:26]),\n",
    "                    \"x\":        float(line[30:38]),\n",
    "                    \"y\":        float(line[38:46]),\n",
    "                    \"z\":        float(line[46:54]),\n",
    "                    \"occ\":      float(line[54:60] or 0),\n",
    "                    \"bfac\":     float(line[60:66] or 0),\n",
    "                })\n",
    "                atom_lines.append(str(line))\n",
    "\n",
    "            # ---- HELIX records ----\n",
    "            elif record == \"HELIX\":\n",
    "                helices.append({\n",
    "                    \"ser_num\":    line[7:10].strip(),       # helix serial number\n",
    "                    \"helix_id\":   line[11:14].strip(),      # helix identifier\n",
    "                    \"start_res\":  line[15:18].strip(),      # initial residue name\n",
    "                    \"start_chain\":line[19].strip(),         # chain identifier\n",
    "                    \"start_seq\":  int(line[21:25]),         # start residue sequence number\n",
    "                    \"end_res\":    line[27:30].strip(),      # ending residue name\n",
    "                    \"end_chain\":  line[31].strip(),         # chain identifier\n",
    "                    \"end_seq\":    int(line[33:37]),         # end residue sequence number\n",
    "                    \"helix_class\":line[38:40].strip(),\n",
    "                    \"length\":     int(line[71:76]),\n",
    "                })\n",
    "                helix_lines.append(str(line))\n",
    "\n",
    "    return atom_lines, helices, helix_lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14e2587f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- planning + header builder to keep total length under a cap (e.g., 200) ---\n",
    "\n",
    "def _hdr_name(name:str) -> str:\n",
    "    return f\"REMARK 999 NAME   {name}\\n\"\n",
    "\n",
    "def _hdr_pdb(pdbid:str) -> str:\n",
    "    return f\"REMARK 999 PDB    {pdbid}\\n\"\n",
    "\n",
    "def _hdr_scaffold(min_len:int, max_len:int) -> str:\n",
    "    # cols 1–16 label; scaffold min at col 20, max at 24\n",
    "    return f\"{'REMARK 999 INPUT':<16}   {min_len:>4}{max_len:>4}\\n\"\n",
    "\n",
    "def _hdr_motif(chain:str, start:int, end:int, group:str='A') -> str:\n",
    "    # cols 1–16 label; chain col 19; start 20–23; end 24–27; group at col 29\n",
    "    return f\"{'REMARK 999 INPUT':<16}  {chain:>1}{start:>4}{end:>4} {group}\\n\"\n",
    "\n",
    "def _hdr_min_total(n:int) -> str:\n",
    "    return f\"{'REMARK 999 MINIMUM TOTAL LENGTH':<31}{'':<6}{n:<3}\\n\"\n",
    "\n",
    "def _hdr_max_total(n:int) -> str:\n",
    "    return f\"{'REMARK 999 MAXIMUM TOTAL LENGTH':<31}{'':<6}{n:<3}\\n\"\n",
    "\n",
    "def build_motif_scaffold_header(name, pdbid, motifs, smin, smax,\n",
    "                                include_leading=False, include_trailing=False):\n",
    "    assert motifs, \"No motifs selected\"\n",
    "    motif_lengths = [end - start + 1 for (_, start, end, _) in motifs]\n",
    "    n_scaff = max(0, len(motifs) - 1)\n",
    "    if include_leading:  n_scaff += 1\n",
    "    if include_trailing: n_scaff += 1\n",
    "    min_total = sum(motif_lengths) + n_scaff * smin\n",
    "    max_total = sum(motif_lengths) + n_scaff * smax\n",
    "\n",
    "    lines = []\n",
    "    lines.append(_hdr_name(name))\n",
    "    lines.append(_hdr_pdb(pdbid))\n",
    "    if include_leading:\n",
    "        lines.append(_hdr_scaffold(smin, smax))\n",
    "    for i, (ch, st, en, grp) in enumerate(motifs):\n",
    "        lines.append(_hdr_motif(ch, st, en, grp))\n",
    "        if i < len(motifs) - 1:\n",
    "            lines.append(_hdr_scaffold(smin, smax))\n",
    "    if include_trailing:\n",
    "        lines.append(_hdr_scaffold(smin, smax))\n",
    "    lines.append(_hdr_min_total(min_total))\n",
    "    lines.append(_hdr_max_total(max_total))\n",
    "    return \"\".join(lines), min_total, max_total\n",
    "\n",
    "def plan_motifs_under_cap(helix_dicts, chain=\"A\", cap_max=200, base_smin=4, base_smax=79):\n",
    "    \"\"\"\n",
    "    Choose a small subset of chain-specific helices and a scaffold range so that MAX_TOTAL ≤ cap_max.\n",
    "    Strategy: prefer shortest helices; try 3→2→1 motifs; if needed, tighten smax (not smin) to fit.\n",
    "    Returns (motifs, smin, smax, min_total, max_total).\n",
    "    \"\"\"\n",
    "    helices_c = [h for h in helix_dicts if h[\"start_chain\"] == chain]\n",
    "    if not helices_c:\n",
    "        # fall back to global shortest helix\n",
    "        h = min(helix_dicts, key=lambda x: (x[\"end_seq\"] - x[\"start_seq\"] + 1))\n",
    "        helices_c = [h]\n",
    "        chain = h[\"start_chain\"]\n",
    "\n",
    "    # sort by helix length ascending\n",
    "    helices_sorted = sorted(helices_c, key=lambda h: (h[\"end_seq\"] - h[\"start_seq\"] + 1))\n",
    "    # try subsets of size 3, then 2, then 1\n",
    "    for k in (3, 2, 1):\n",
    "        if len(helices_sorted) < k:\n",
    "            continue\n",
    "        cand = helices_sorted[:k]\n",
    "        motifs = [(chain, h[\"start_seq\"], h[\"end_seq\"], \"A\") for h in cand]\n",
    "        motif_sum = sum(h[\"end_seq\"] - h[\"start_seq\"] + 1 for h in cand)\n",
    "        n_scaff = max(0, k - 1)  # scaffolds BETWEEN motifs only\n",
    "\n",
    "        # start with base scaffold range\n",
    "        smin, smax = base_smin, base_smax\n",
    "        min_total = motif_sum + n_scaff * smin\n",
    "        max_total = motif_sum + n_scaff * smax\n",
    "\n",
    "        if min_total > cap_max:\n",
    "            # even the minimum can't fit; try fewer motifs\n",
    "            continue\n",
    "\n",
    "        if max_total <= cap_max:\n",
    "            return motifs, smin, smax, min_total, max_total\n",
    "\n",
    "        # tighten smax just enough to fit\n",
    "        if n_scaff > 0:\n",
    "            needed_smax = (cap_max - motif_sum) // n_scaff\n",
    "            smax_adj = max(smin, min(smax, needed_smax))\n",
    "            max_total_adj = motif_sum + n_scaff * smax_adj\n",
    "            if max_total_adj <= cap_max:\n",
    "                return motifs, smin, smax_adj, min_total, max_total_adj\n",
    "\n",
    "        # else try fewer motifs\n",
    "    # If nothing fits (very long helices), fall back to the single shortest helix, no scaffold\n",
    "    h = helices_sorted[0]\n",
    "    motifs = [(chain, h[\"start_seq\"], h[\"end_seq\"], \"A\")]\n",
    "    smin = smax = base_smin\n",
    "    min_total = max_total = (h[\"end_seq\"] - h[\"start_seq\"] + 1)\n",
    "    return motifs, smin, smax, min_total, max_total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "01508803",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4wav [('A', 192, 200, 'A'), ('A', 172, 191, 'A'), ('A', 16, 38, 'A')] (4, 74) 60 200\n",
      "4fbz [('A', 232, 234, 'A'), ('A', 192, 195, 'A'), ('A', 226, 231, 'A')] (4, 79) 21 171\n",
      "4bem [('A', 2, 19, 'A'), ('A', 22, 43, 'A'), ('A', 46, 78, 'A')] (4, 63) 81 199\n",
      "4tsy [('A', 129, 139, 'A'), ('A', 5, 29, 'A')] (4, 79) 40 115\n",
      "5cbg [('A', 6, 12, 'A'), ('A', 49, 60, 'A'), ('A', 26, 45, 'A')] (4, 79) 47 197\n",
      "2oar [('A', 1, 11, 'A'), ('A', 90, 100, 'A'), ('A', 105, 124, 'A')] (4, 79) 50 200\n",
      "4wab [('A', 54, 59, 'A'), ('A', 46, 53, 'A'), ('A', 154, 161, 'A')] (4, 79) 30 180\n",
      "4rng [('A', 23, 34, 'A'), ('A', 6, 23, 'A'), ('A', 40, 61, 'A')] (4, 74) 60 200\n",
      "5a43 [('A', 28, 30, 'A'), ('A', 72, 78, 'A'), ('A', 62, 70, 'A')] (4, 79) 27 177\n",
      "4hyj [('A', 33, 36, 'A'), ('A', 70, 74, 'A'), ('A', 95, 102, 'A')] (4, 79) 25 175\n",
      "5ax0 [('A', 67, 69, 'A'), ('A', 167, 169, 'A'), ('A', 162, 165, 'A')] (4, 79) 18 168\n",
      "3am6 [('A', 32, 34, 'A'), ('A', 179, 188, 'A'), ('A', 195, 211, 'A')] (4, 79) 38 188\n",
      "4pop [('A', 38, 40, 'A'), ('A', 134, 137, 'A'), ('A', 174, 178, 'A')] (4, 79) 20 170\n",
      "4qnc [('A', 16, 28, 'A'), ('A', 1, 16, 'A'), ('A', 34, 56, 'A')] (4, 74) 60 200\n",
      "4qnd [('A', 37, 44, 'A'), ('A', 26, 35, 'A'), ('A', 1, 12, 'A')] (4, 79) 38 188\n",
      "4xu4 [('A', 152, 155, 'A'), ('A', 47, 65, 'A'), ('A', 100, 120, 'A')] (4, 78) 52 200\n"
     ]
    }
   ],
   "source": [
    "# now retrieve residue numbers and atomic coords\n",
    "folder = \"alpha_helix_pdbs\"\n",
    "for id in good_pdbids:\n",
    "    fname = f\"{folder}/{id}.pdb\"\n",
    "    atoms, helices, helix_lines = parse_pdb(fname)\n",
    "    out = []\n",
    "    for i in range(len(helices)):\n",
    "        row = {\n",
    "            \"helix_number\" : helices[i][\"helix_id\"],\n",
    "            \"start_res_number\": helices[i][\"start_seq\"],\n",
    "            \"end_res_number\": helices[i][\"end_seq\"],\n",
    "            \"start_res_code\": helices[i][\"start_res\"],\n",
    "            \"end_res_code\": helices[i][\"end_res\"],\n",
    "            \"chain_identifier\": helices[i][\"end_chain\"],\n",
    "            \"helix_class\" : helices[i][\"helix_class\"],\n",
    "            \"length\" : helices[i][\"length\"]\n",
    "            }\n",
    "        out.append(row)\n",
    "    df_temp = pd.DataFrame(out)\n",
    "    df_temp.to_csv(f\"alpha_helix_residue_numbers/{id}.csv\", index=False)\n",
    "    \n",
    "    # now process atomic coords\n",
    "    atom_file_name = f\"alpha_helix_atomic_coords/{id}_atoms.pdb\"\n",
    "    with open(atom_file_name, 'w') as f:\n",
    "        for line in atoms:\n",
    "            f.write(line)\n",
    "    # now process helix coords/lines\n",
    "    helix_file_name = f\"alpha_helix_pdbs_helix_only/{id}_helix.pdb\"\n",
    "    with open(helix_file_name, 'w') as f:\n",
    "        for line in helix_lines:\n",
    "            f.write(line)\n",
    "    # write combo with header \n",
    "    helix_file_name = f\"alpha_helix_templates/{id}_helix.pdb\"\n",
    "    with open(helix_file_name, 'w') as f:\n",
    "        # write header\n",
    "        # write combo with header \n",
    "        helix_file_name = f\"alpha_helix_templates/{id}_helix.pdb\"\n",
    "        with open(helix_file_name, 'w') as f:\n",
    "            # plan motifs + scaffolds to keep max ≤ 200\n",
    "            motifs, smin, smax, min_tot, max_tot = plan_motifs_under_cap(\n",
    "                helices, chain=\"A\", cap_max=200, base_smin=4, base_smax=79\n",
    "            )\n",
    "\n",
    "            header, min_total, max_total = build_motif_scaffold_header(\n",
    "                name=f\"guided_{id}\",\n",
    "                pdbid=id,\n",
    "                motifs=motifs,\n",
    "                smin=smin,\n",
    "                smax=smax,\n",
    "                include_leading=False,\n",
    "                include_trailing=False,\n",
    "            )\n",
    "            # optional: quick sanity log\n",
    "            print(id, motifs, (smin, smax), min_total, max_total)\n",
    "\n",
    "            # write header then HELIX+ATOM\n",
    "            f.write(header + \"\\n\")\n",
    "            for line in helix_lines:\n",
    "                f.write(line)\n",
    "            for line in atoms:\n",
    "                f.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3533f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# folder = \"alpha_helix_pdbs\"\n",
    "# for id in good_pdbids:\n",
    "#     fname = f\"{folder}/{id}.pdb\"\n",
    "#     atoms, helices, helix_lines = parse_pdb(fname)  # helices is defined here!\n",
    "\n",
    "#     # --- residue CSV export ---\n",
    "#     out = []\n",
    "#     for h in helices:\n",
    "#         row = {\n",
    "#             \"helix_number\": h[\"helix_id\"],\n",
    "#             \"start_res_number\": h[\"start_seq\"],\n",
    "#             \"end_res_number\": h[\"end_seq\"],\n",
    "#             \"start_res_code\": h[\"start_res\"],\n",
    "#             \"end_res_code\": h[\"end_res\"],\n",
    "#             \"chain_identifier\": h[\"end_chain\"],\n",
    "#             \"helix_class\": h[\"helix_class\"],\n",
    "#             \"length\": h[\"length\"],\n",
    "#         }\n",
    "#         out.append(row)\n",
    "#     df_temp = pd.DataFrame(out)\n",
    "#     df_temp.to_csv(f\"alpha_helix_residue_numbers/{id}.csv\", index=False)\n",
    "\n",
    "#     # --- ATOM-only PDB ---\n",
    "#     atom_file_name = f\"alpha_helix_atomic_coords/{id}_atoms.pdb\"\n",
    "#     with open(atom_file_name, \"w\") as f:\n",
    "#         for line in atoms:\n",
    "#             f.write(line)\n",
    "\n",
    "#     # --- HELIX-only PDB ---\n",
    "#     helix_file_name = f\"alpha_helix_pdbs_helix_only/{id}_helix.pdb\"\n",
    "#     with open(helix_file_name, \"w\") as f:\n",
    "#         for line in helix_lines:\n",
    "#             f.write(line)\n",
    "\n",
    "#     # --- Template with REMARK 999 header + HELIX + ATOM ---\n",
    "#     helix_file_name = f\"alpha_helix_templates/{id}_helix.pdb\"\n",
    "#     with open(helix_file_name, \"w\") as f:\n",
    "#         # select motifs (e.g., three short chain A helices)\n",
    "#         helices_A = [h for h in helices if h[\"start_chain\"] == \"A\"]\n",
    "\n",
    "#         preferred = {(1, 11), (90, 100), (105, 124)}\n",
    "#         motifs = [\n",
    "#             (\"A\", h[\"start_seq\"], h[\"end_seq\"], \"A\")\n",
    "#             for h in helices_A\n",
    "#             if (h[\"start_seq\"], h[\"end_seq\"]) in preferred\n",
    "#         ]\n",
    "\n",
    "#         if len(motifs) < 3 and helices_A:\n",
    "#             remaining = [h for h in helices_A if (h[\"start_seq\"], h[\"end_seq\"]) not in preferred]\n",
    "#             remaining = sorted(remaining, key=lambda x: (x[\"end_seq\"] - x[\"start_seq\"] + 1))\n",
    "#             for h in remaining[: 3 - len(motifs)]:\n",
    "#                 motifs.append((\"A\", h[\"start_seq\"], h[\"end_seq\"], \"A\"))\n",
    "\n",
    "#         if not motifs and helices:\n",
    "#             h = min(helices, key=lambda x: (x[\"end_seq\"] - x[\"start_seq\"] + 1))\n",
    "#             motifs = [(h[\"start_chain\"], h[\"start_seq\"], h[\"end_seq\"], \"A\")]\n",
    "\n",
    "#         header = build_motif_scaffold_header(\n",
    "#             name=f\"helices_small_total_{id}\",\n",
    "#             pdbid=id,\n",
    "#             motifs=motifs,\n",
    "#             scaffold_range=(4, 79),\n",
    "#             include_leading_scaffold=False,\n",
    "#             include_trailing_scaffold=False,\n",
    "#         )\n",
    "\n",
    "#         # write header, then HELIX lines, then ATOM lines\n",
    "#         f.write(header + \"\\n\")\n",
    "#         for line in helix_lines:\n",
    "#             f.write(line)\n",
    "#         for line in atoms:\n",
    "#             f.write(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e935a4c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esmfold_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
